### PR TITLE
Update Subscriptions with WooPayments eligibility as we move to deprecate this functionality

### LINKED ISSUE #6510: Deprecate subscriptions functionality bundled in WC Payments
### Description

As part of our project to sunset the subscriptions functionality bundled in WC Payments we will be deprecating that feature.

This issue will track the changes necessary to handle that deprecation. 

### Acceptance criteria

- [x] Do not load subscription functionality on new sites created after the release of the version of WC Pay containing these changes.
- [x] No longer load subscription functionality on existing WC Pay sites when:
    1. Store has no active Stripe Billing Subscriptions; and
    2. Store has no published subscription products in their catalog.

### Additional context

peNR48-2u-p2 project.

### PR DESCRIPTION
**Pull request by james-allan**
> [!NOTE]  
> This PR reintroduces the changes made in #7117 that were reverted in #7194 when it was decided that deprecating the WCPay Subscriptions feature would come in the next release. 

Fixes #6510 

#### Changes proposed in this Pull Request

As part of the [Subscriptions Primary Offering](https://github.com/Automattic/woocommerce-payments/milestone/34) project, we need to deprecate the existing WCPay Subscriptions feature so that merchants can opt into the new subscriptions setting.

In this PR we are updating the eligibility of the WCPay Subscriptions so that we:
 - Do not load the **Advanced > WCPay Subscription** setting on new sites created after the release of the version of WC Pay containing these changes.
 - No longer load subscription functionality on existing sites that have no active Stripe Billing Subscriptions or published subscription products.

To meet these new requirements I've changed `is_wcpay_subscriptions_eligible()` to only return true when:

1. Store located in the US
2. Doesn't have the WC Subscriptions extension active
3. Has either one of the following:
   1. the store has existing WCPay Subscriptions or,
   2. has published subscription products

#### Testing instructions

Testing this PR requires toggling the following:
 - Deactivate/activate WooCommerce Subscriptions
 - Update the store's base country to US and non-US
 - Has no published subscription products
 - Has no WCPay subscriptions (with `_wcpay_subscription_id` metadata)

With that information, here are the testing instructions to toggle through most of the new eligibility cases:

1. Setup the state of your store so that you have 1 subscription product, 1 WCPay Subscription, WC Subscriptions deactivated, and the store is based in the US
2. Confirm the **Enable Subscriptions with WooPayments** setting is found in **Payments > Settings > Advanced**
3. Activate Subscriptions, confirm the setting is gone
4. Deactivate Subscriptions to bring it back
5. Change the stores base country to a non-US country, confirm the WCPay Subscriptions setting is gone
6. Change the country back to the US
7. Set your subscriptions product to private and modify the `_wcpay_subscription_id` to something like `_wcpay_subscription_id_test`
8. Confirm the setting is no longer shown
9. Set the subscription product back to published, confirm the setting is back.
10. Put the product back to private and change the metadata on the subscription to bring back the `_wcpay_subscription_id` meta.
11. Confirm the WCPay subscription setting is loading.

With these instructions I have tested the following cases:

- **Setting doesn't show if any of the following conditions are met:**
   - [x] Has WC Subscriptions active
   - [x] Non-US based store
   - [x] Doesn't have existing subscriptions products
   - [x] Doesn't have any subscriptions with `__wcpay_subscriptions_id` meta (including pending, canceled, on-hold etc.)
- **Settings only shows when:**
   - [x] No WC Subscriptions active, country set to US, no subscriptions but has existing products
   - [x] No WC Subscriptions active, country set to US, has existing subscriptions 

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7238` or branch name `issue/6510-deprecate-wcpay-subscriptions` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=issue/6510-deprecate-wcpay-subscriptions) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 385e2f8d7ff5669b7899b0ebc0abda8448cb6b45
- Build time: 2023-10-06 02:38:03 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: includes/subscriptions/class-wc-payments-subscriptions.php [hunk 1]
```diff
@@ -157,4 +159,15 @@ public static function is_duplicate_site() {
...
 		return class_exists( 'WCS_Staging' ) && WCS_Staging::is_duplicate_site();
 	}
+
+	/**
+	 * Disable the WCPay Subscriptions feature on WooPayments plugin update if it's enabled and the store is no longer eligible.
+	 *
+	 * @see WC_Payments_Features::is_wcpay_subscriptions_eligible() for eligibility criteria.
+	 */
+	public static function maybe_disable_wcpay_subscriptions_on_update() {
+		if ( WC_Payments_Features::is_wcpay_subscriptions_enabled() && ! WC_Payments_Features::is_wcpay_subscriptions_eligible() ) {
```
Comment (james-allan) [inline includes/subscriptions/class-wc-payments-subscriptions.php]: I wonder if we need to change this to also disable this feature if the Woo Subscriptions plugin is active.

Imagine there's a site that has used WCPay Subscriptions in the past, installed Woo Subscriptions, if they deactivate Woo Subscriptions do we want them to automatically switch to using WCPay Subscriptions feature?

Comment (james-allan) [reply includes/subscriptions/class-wc-payments-subscriptions.php]: In line with the discussion here (p1695182789067919/1695179380.867469-slack-C055WHLA98D), Brent agrees and we'll disable the WCPay Subscriptions feature if its enabled and the Woo Subscriptions plugin is already active at the time of next WooPayments update.

**Comment by james-allan** [comment]
I ran through the following scenarios when testing this PR:

- [x] **WCPay subscriptions previously enabled**
   - [x] Disabled for stores with no products or subscriptions.
   - [x] Left enabled for stores with at least 1 product. 
   - [x] Left enabled for stores with at least 1 subscription. 
   - [x] Disabled on sites with at least 1 Stripe Billing Subscription but Woo Subscriptions 
enabled.
   - [x] Disabled on sites with no Stripe Billing Subscriptions but Woo Subscriptions enabled.
- [x] **WCPay subscriptions previously disabled** - _no change_.

In terms of the impact of these changes. The result should be:

- If you weren't using it - no change. 
- If it is enabled: 
    - But you have the Woo Subscriptions plugin - it's **disabled**. Any existing Stripe Billing subscriptions can be served by Woo Subscriptions. Disabling Woo Subscriptions won't restore WCPay Subscriptions functionality. Doing so would have disproportionate impacts on the site. Disabling Woo Subscriptions while you have Stripe Billing subscriptions will still display an appropriate notice. See https://github.com/Automattic/woocommerce-payments/issues/7240
    - You don't have Woo Subscriptions and you don't have WCPay Subscriptions - its **disabled**. The site wasn't using it. 
    - You don't have Woo Subscriptions and do have WCPay Subscriptions - no change (it's left enabled).

I'll need to see what impact this has on GlobalStep testing. I suspect this change might disable this feature preventing them from testing WCPay subscription flows in the critical flows.

**Comment by james-allan** [comment]
> I'll need to see what impact this has on GlobalStep testing. I suspect this change might disable this feature preventing them from testing WCPay subscription flows in the critical flows.

The critical flows testing instructions specially asks the tester to enable the WCPay subscriptions feature via the dev tools plugin: https://github.com/Automattic/woocommerce-payments/wiki/WC-Pay-Subscriptions-critical-flows-testing-instructions#enable-wc-pay-subscriptions

This means that even though this setting is being removed and deprecated Global step can still access this feature for testing via dev tools.

**Review by mattallan — COMMENTED** [review]
In [0e8dfb6](https://github.com/Automattic/woocommerce-payments/pull/7238/commits/0e8dfb6538292f169cbeae31601680a8f596faef) I've pushed up a change which moves the disabling of the legacy WCPay Subscriptions feature to the main payments class to fix an issue where if you had the legacy wcpay subscription feature enabled & Woo Subscriptions activated with no Stripe Billing subscriptions, the feature was not being disabled.

This was because this code was inside a class that was no longer loading.

On the latest changes I've ran through the following checks:

- [x] WCPay subscriptions previously enabled
  - [x] Disabled for stores with no products or subscriptions.
  - [x] Left enabled for stores with at least 1 product.
  - [x] Left enabled for stores with at least 1 subscription.
  - [x] Disabled on sites with at least 1 Stripe Billing Subscription but Woo Subscriptions enabled.
- [x] Disabled on sites with no Stripe Billing Subscriptions but Woo Subscriptions enabled.
- [x] WCPay subscriptions previously disabled - no change.

---

Worth noting we also discussed hooking onto the Woo Subscriptions plugin activation and disabling the WCPay Subscriptions feature there as well however we couldn't find a clean way to hook onto that event so we're going to leave it.

**Review by mattallan — APPROVED** [review]
@james-allan as we discussed, I've moved the code which used the WooPayments updated hook to now be inside the main payments class and ran through all of the checks again to confirm we're turning off the legacy WCPay Subscriptions feature correctly.

Approving this to be merged.

**Pull request merged**